src/ints/mouse_dos_driver.cpp

```diff
@@ -0,0 +1,1440 @@
+}
+
+bool MOUSEDOS_NotifyPressed(const MouseButtons12S new_buttons_12S,
+                            const uint8_t idx, const MouseEventId event_id)
+{
+    assert(idx < NUM_BUTTONS);
+
+    buttons = new_buttons_12S;
+
+    state.times_pressed[idx]++;
+    state.last_pressed_x[idx] = GetPosX();
+    state.last_pressed_y[idx] = GetPosY();
+
+    return MOUSEDOS_HasCallback(static_cast<uint8_t>(event_id));
+}
+
+bool MOUSEDOS_NotifyReleased(const MouseButtons12S new_buttons_12S,
+                             const uint8_t idx, const MouseEventId event_id)
+{
+    assert(idx < NUM_BUTTONS);
+
+    buttons = new_buttons_12S;
+
+    state.times_released[idx]++;
+    state.last_released_x[idx] = GetPosX();
+    state.last_released_y[idx] = GetPosY();
+
+    return MOUSEDOS_HasCallback(static_cast<uint8_t>(event_id));
+}
+
+bool MOUSEDOS_NotifyWheel(const int16_t w_rel)
+{
+    if (!state.cute_mouse) // wheel only available if CuteMouse extensions
+                           // are active
+        return false;
+
+    const auto tmp = std::clamp(static_cast<int32_t>(w_rel + wheel),
+                                static_cast<int32_t>(INT16_MIN),
+                                static_cast<int32_t>(INT16_MAX));
+
+    wheel                    = static_cast<int16_t>(tmp);
+    state.last_wheel_moved_x = GetPosX();
+    state.last_wheel_moved_y = GetPosY();
+
+    return MOUSEDOS_HasCallback(static_cast<uint8_t>(MouseEventId::WheelHasMoved));
+}
+
+static Bitu INT33_Handler()
+{
+    switch (reg_ax) {
+    case 0x00: // MS MOUSE - reset driver and read status
+        ResetHardware();
+        [[fallthrough]];
+    case 0x21:               // MS MOUSE v6.0+ - software reset
+        reg_ax = 0xffff; // mouse driver installed
+        reg_bx = 3;      // for 2 buttons return 0xffff
+        Reset();
+        break;
+    case 0x01: // MS MOUSE v1.0+ - show mouse cursor
+        if (state.hidden)
+            state.hidden--;
+        state.update_region_y[1] = -1; // offscreen
+        MOUSEDOS_DrawCursor();
+        break;
+    case 0x02: // MS MOUSE v1.0+ - hide mouse cursor
+        if (CurMode->type != M_TEXT)
+            RestoreCursorBackground();
+        else
+            RestoreCursorBackgroundText();
+        state.hidden++;
+        break;
+    case 0x03: // MS MOUSE v1.0+ / CuteMouse - get position and button status
+        reg_bl = buttons.data;
+        reg_bh = GetResetWheel8bit(); // CuteMouse clears wheel counter too
+        reg_cx = GetPosX();
+        reg_dx = GetPosY();
+        break;
+    case 0x04: // MS MOUSE v1.0+ - position mouse cursor
+    {
+        // If position isn't different from current position, don't
+        // change it. (position is rounded so numbers get lost when the
+        // rounded number is set) (arena/simulation Wolf)
+        if ((int16_t)reg_cx != GetPosX())
+            pos_x = static_cast<float>(reg_cx);
+        if ((int16_t)reg_dx != GetPosY())
```

Comment:
Use static_cast in the ifs too?

Reply:
signed<=>unsigned casting causes warnings with some compiler - fixed using helper function